### ISSUE TITLE
Should we deprecate the existing Cross Resource References on Managed Resources?

### ISSUE THREAD

**Issue by turkenh** [crossplane/crossplane#4627]
### What problem are you facing?

Crossplane Managed Resources are building blocks representing external API resources (typically cloud resources) on a control plane. It is a common need to derive one field of a resource from another one, i.e. a `vpcId` in the spec of an AWS `Subnet` resource. In the early days of Crossplane, a concept known as [Cross Resource Referencing](https://github.com/crossplane/crossplane/pull/874) was introduced to solve this problem. 
See the following as an example:

```yaml
apiVersion: ec2.aws.upbound.io/v1beta1
kind: VPC
metadata:
  name: sample-vpc
spec:
  ...
---
apiVersion: ec2.aws.upbound.io/v1beta1
kind: Subnet
metadata:
  name: sample-subnet1
spec:
  forProvider:
    ...
    vpcIdRef:
      name: sample-vpc
```

The provider will resolve the `vpcIdRef` field in this example once the VPC is ready. This approach has been working pretty well and widely adopted in the community; however, it has with the following [limitations](https://github.com/crossplane/crossplane/pull/4366#discussion_r1286504226):

> - Requires all possible references to be manually identified and declared at build time.
> - Is inflexible - users can only reference types the provider developers declared, in the ways they declared.
> - Requires the referring provider to import any types they refer to (in the case of family providers).

With the invention of the Compositions, which is a layer on top of the MRs, there are new ways to `patch` fields of composed MRs either from the Claims/Composite or from other MRs with the help of `ToComposite/FromComposite` or `ToEnvironment/FromEnvironment` type patches. The solution for patching between MRs is still not ideal (requires two patches); however, Composition functions could provide a much better experience here by enabling the usage of general-purpose languages or with the help of more powerful composers (e.g. [Go Template Function](https://github.com/crossplane/crossplane/issues/4573)).

During the discussions on the [Generic Cross Resource Reference PR](https://github.com/crossplane/crossplane/pull/4366), we realized that we are not willing to add more functionality/complexity at the MR level, especially considering the similar functionalities we have at the composition layer. 

From an architectural point of view, it would be cleaner if we could keep the primitives (Managed Resources) as simple as possible, just doing the job they are for, e.g. representing the external resource on the control plane, and we push any other functionality to upper layers. 

### How could Crossplane help solve your problem?

Discuss whether we want to see the MR-level Cross Resource References in the future of Crossplane or if we should start considering deprecating them. We know they are widely used in the wild, and some users do not use Composition at all today, so please don't consider this issue something to be done in the short-mid term but more like a conversation starter to collect some ideas.

**Comment by ytsarev** [comment]
My first reaction is definitely skeptical :) 

* It will break tons of existing Compositions. The migration to the new reference style will be non-trivial. 
* Crossplane users will lose the ability to quickly prototype a solution without Compositions, e.g. with a raw yaml sequence of MRs. It proved to be a very practical approach.
* Crossplane users will lose the ability to use providers/MRs without Compositions like templating with Helm(controversial but used in the wild,; the need will go away with Functions, but existing solutions are at risk)

**Comment by elohmrow** [comment]
_philosophically_, deprecating (however it is defined) _makes sense_.
_practically_, deprecating (however it is defined) _will be massively problematic for existing solutions_.

if deprecating (however it is defined) occurs, please make sure:
  - notice is served well in advance (a year is not unreasonable for such a big change)
  - useful / clear logs are emitted 
  - there is a clear alternative, with low cognitive load, that is easy to use, and ready to use, before deprecating (however it is defined)

**Comment by negz** [comment]
For future readers, especially folks implementing new providers, I'd recommend not implementing cross resource references for the reasons @turkenh outlines above.

https://github.com/crossplane/crossplane/blob/v1.16.0/design/design-doc-composition-functions-extra-resources.md
https://github.com/crossplane-contrib/function-extra-resources/tree/main

Instead, I recommend using composition functions. A functions-based composition can request extra resources - resources that aren't the XR or any of its composed resources. It can then use any function (including your own logic written in Go/Python/etc) to derive XR or composed resource values from the extra resource.

Generally this is much more flexible/powerful than cross-resource references.

**Comment by yordis** [comment]
Cross-resource reference is one of the most critical features of Crossplane from my perspective ...

**Comment by p4ck3t0** [comment]
Does that mean to create a VM in a VPC for example in GCP one should use a composition?
For that Scenario, one needs to create the VPC first and then place the VM within the created VPC. Is there a already existing composition for it? If not, one needs to create a composition for only that little use case?

I might get something wrong, but that seems to be a lot of effort, just to spin up a VM and a VPC.

**Comment by negz** [comment]
Yes, I would recommend using compositions.

**Comment by p4ck3t0** [comment]
Hm oke I see. Will there be compositions available for things like creating a virtual machine, because one needs three resources for that at least in azure? Or does one need to create his own compositions for everything?

For me it feels like, that the complexity has rapidly improved with compositions and it feels very overengineered, for smaller use cases, like just spawning a vm (Virtualmachine, NetworkInterface, PublicIP). Creating a composition for this use case kind of doesn't feels right. The resource references work beautifully for that use case. In addition, I don't know if composition can be referenced. E.g. if I create two composition:

1. Virtual Machine Composition (VM, Networkinterface, Public IP)
2. VPC with Subnet (VPC, Subnet)

And now want to deploy a VM into a VPC, do I need to create a third composition with both resources or can composition 1 reference to composition 2? The Go templating approach seems promising, but (I think) lacks conditions.

But I can see and understand the decision to move the resource dependency management to an upper layer. Also for complex environments with many resources its an important feature that makes management and reproducibility easier.

What I will probably do, is write my own resource dependency resolver once crossplane deprecates the resource referencing.

**Comment by Rand0mF** [comment]
I agree with @p4ck3t0. There are many "small" usecases out there with just a few MRs. Everytime creating a full blown composition with XRD introduces high complexity. 

And I agree with @ytsarev, it would break everything out there right now. From my experience, I can see helm commonly used for crossplane resources and I would even encourage it. Helm is the tool when it comes to kubernetes resource bundling. When we deprecate cross resource references, we will lock out helm for crossplane usage which will have a major impact for adaption. Rather I'd suggest to turn 180° from this proposal and encourage cross resource references and even add more things to make tools like helm more useable, like [generic cross-references](https://github.com/crossplane/crossplane/pull/4366). Let the users decide which tool to use. Compositions are great but don't force them and lock out all other tooling.

**Comment by sdischer-sap** [comment]
I do agree with the notion that the provider should generally just be another interface to interacting with the underlying api and should not added assumptions or opinionated logic on top of that. However relating entities within the scope of one provider to each does not feel like a violation to this premise to me and feels like a core feature that shouldn't need to be solved by each enduser on its own. Specifically since it would massively raise the learning curve for newcomers until they would achieve anything useful. I would be afraid to loss a lot of the excitement with them along the way if they were to "implement" those connections on their own.

**Comment by negz** [comment]
Per my last comment I think we have a decision: no we won't drop them. We'll avoid adding more functionality, though.
